### PR TITLE
Adds some Documentation

### LINKED ISSUE #2: Documentation
Documentation needs to be written with good examples

### PR DESCRIPTION
**Pull request by meiamsome**
Fixes #2

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -0,0 +1,207 @@
+# graphql-auto-requester
+
+`graphql-auto-requester` is a tool for making working with GraphQL requests a simpler and more efficient by only
```
Comment (steve7411) [inline Readme.md]: A simpler and more efficient what?

File: Readme.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  sub(input: Int!): Num!
+}
+
+type Query {
+  getNumber(input: Int!): Num!
+}
+```
+
+You would usually have to manually define your query, which could get complex depending upon your use case. Using the
+requester we are able to traverse the schema and onlty request data when it is required. For example:
```
Comment (animalcharade) [inline Readme.md]:
```suggestion
requester we can traverse the schema and only request data when it is required. For example:
```

File: Readme.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
+# graphql-auto-requester
+
+`graphql-auto-requester` is a tool for making working with GraphQL requests a simpler and more efficient by only
```
Comment (animalcharade) [inline Readme.md]:
```suggestion
`graphql-auto-requester` is a tool for making working with GraphQL requests simpler and more efficient by only
```

File: Readme.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    add(input: 10) {
+      mult(input: 5) {
+        value
+      }
+    }
+  }
+}
+```
+
+However, it is hard for us to implement a feature that explores the graph very deeply without having to struggle to
```
Comment (animalcharade) [inline Readme.md]:
```suggestion
However, it is hard for us to implement a feature that explores the graph very deeply without struggling to
```

File: Readme.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+maintain one query file that represents the superset of all options we may want to perform. This also means that when we
```
Comment (animalcharade) [inline Readme.md]:
```suggestion
maintain one query file that represents the superset of all options we may want to perform. This also means that, when we
```

File: Readme.md [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+  sub(input: Int!): Num!
+}
+
+type Query {
+  getNumber(input: Int!): Num!
+}
+```
+
+You would usually have to manually define your query, which could get complex depending upon your use case. Using the
+requester we are able to traverse the schema and onlty request data when it is required. For example:
```
Comment (steve7411) [inline Readme.md]: Missing a comma after the introductory phrase

Comment (steve7411) [inline Readme.md]: only is misspelled

File: Readme.md [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+```js
+const requester = new GraphQLAutoRequester(schema)
+const result = await requester.query.getNumber({ input: 100 })
+  .add({ input: 10 })
+  .mult({ input: 5 })
+  .value
+console.log(result) // 550
+```
+
+This is equivalent to the following GraphQL query (And this is the query that would be automatically executed)
```
Comment (steve7411) [inline Readme.md]: And should not be capitalized

File: Readme.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
```
Comment (animalcharade) [inline Readme.md]:
```suggestion
This is equivalent to the following GraphQL query, which would be automatically executed:
```

File: Readme.md [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+  } else {
+    return 1 + await collatzByQuery(result.add.value)
+  }
+}
+
+// Calling example
+const steps = await collatzByQuery(100)
+console.log(steps) // 10
+```
+There are some problems with this implementation, mainly that we are requesting the result of the division and of the
```
Comment (animalcharade) [inline Readme.md]:
```suggestion
There are some problems with this implementation, mainly that we are requesting the results of both the division and the
```

File: Readme.md [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+// Calling example
+const requester = new GraphQLAutoRequester(schema)
+const steps = await collatzAutoRequester(requester.getNumber({ input: 100 }))
+console.log(steps) // 10
+```
+
+### Query Aggregation
+That may not seem like that dramatic a change, but it is when we start running queries in parallel that we get to use
```
Comment (animalcharade) [inline Readme.md]:
```suggestion
That may not seem like a dramatic change, but when we start running queries in parallel we get to use
```

File: Readme.md [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+  } else {
+    return 1 + await collatzByQuery(result.add.value)
+  }
+}
+
+// Calling example
+const steps = await collatzByQuery(100)
+console.log(steps) // 10
+```
+There are some problems with this implementation, mainly that we are requesting the result of the division and of the
```
Comment (steve7411) [inline Readme.md]: result should be plural

**Review by animalcharade — COMMENTED** [review]
Grammar things

File: Readme.md [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+// Calling example
+const requester = new GraphQLAutoRequester(schema)
+const steps = await collatzAutoRequester(requester.getNumber({ input: 100 }))
+console.log(steps) // 10
+```
+
+### Query Aggregation
+That may not seem like that dramatic a change, but it is when we start running queries in parallel that we get to use
+the full power of `graphql-auto-requester`. Fields that are requested in the same tick of the JS event loop will be
+bundled in to a single request.
```
Comment (steve7411) [inline Readme.md]:
```suggestion
That may not seem like too dramatic a change, but it is when we start running queries in parallel; then we get to use
the full power of `graphql-auto-requester`. Fields that are requested in the same tick of the JS event loop will be
bundled in to a single request.
```

File: Readme.md [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+        value
+      }
+    }
+  }
+}
+```
+We would see 3 queries to the upstream GraphQL, as the mod and square branches are executed in parallel by the GraphQL
+executor on our new service.
+Importantly, the request for `__typename` does NOT result in a further request to the upstream service, as we can tell
+from the Schema that this is valid and do not need to execute a request for any subfields as `__typename` is simply
```
Comment (steve7411) [inline Readme.md]: Why is schema capitalized here?

File: src/index.ts [hunk 14]
```diff
@@ -6,18 +6,17 @@ import { createResultProxy } from 'graphql-result-proxy'
 import AutoGraphQLObjectType from './ObjectType'
 import { lazyProperty } from './utils'
 
+export { default as AutoGraphQLObjectType } from './ObjectType'
+
 export default class GraphQLAutoRequester {
   schema: GraphQLSchema
-  // TODO: This should not live here
```
Comment (steve7411) [inline src/index.ts:11]: How dare you!

**Review by steve7411 — CHANGES_REQUESTED** [review]

**Review by animalcharade — APPROVED** [review]

**Review by steve7411 — APPROVED** [review]

**Pull request merged**